ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -101,6 +103,13 @@ typedef struct {
   live_object_data object_data;
 } partial_heap_recording;
 
+typedef struct {
+  // Has ownership of this, needs to clean-it-up if not transferred.
+  heap_record *heap_record;
+  long obj_id;
+  live_object_data object_data;
+} uncommitted_sample;
```

Comment:
To be honest, I've been starting at this PR for quite a while and thinking... is there a way we can avoid the additional complexity of the queue?

I'm starting to think that maybe there's an option. Looking at the implementation of `rb_st_copy` (which calls `st_replace`) and it seems like it would be blazing fast -- it has 2 malloc and 2 memcopys and it's done.

What if, instead of having to deal with the lock, we created a copy of the `object_records` at the end of `heap_recorder_flush` and then used that copy when calling `heap_recorder_for_each_live_object`?

This is all a bit pie in the sky, but the tradeoff of one copy of the `object_records` per flush would be no locking needed + no queueing needed.

That said, I may be missing something so I'll give a pass on this PR still and leave the final decision to you.

Reply:
Hmm yeah you may be onto something here. I had assumed the map copy would be slower but it seems to be quite fast indeed. I'll follow your suggestion and maybe use this as a benchmarking case to move our pie closer to the ground 😂

Reply:
Implemented!